product/specs/2022-01-views/02-modeling-views.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Views: Modeling Views
+description:
+published: true
+date: 2022-01-13T19:49:54Z
+tags: 
+editor: markdown
+dateCreated: 2022-01-13T19:49:54Z
+---
+
+Here's how I think we should model views in our API and UI. Each heading represents an attribute of Views.
+
+Before I get into attributes, here's an example table named `Movies` that I'll use to illustrate different types of Views.
+
+| ID | Title | Year | Genre |
+|-|-|-|-|
+| 1 | The River Wild | 1994 | Thriller |
+| 2 | Don't Look Up | 2021 | Comedy |
+| 3 | Daylight | 1996 | Action |
+| 4 | Jason Bourne | 2016 | Action |
+
+## Query
+This is the SQL query that defines the view. This is **required**.
+
+## Columns
+Output columns of the View's Query, they will be shown in the UI.  At least one column is **required** for a view.
+
+Column attributes are defined [in a separate page](/product/specs/2022-01-views/03-modeling-view-columns.md).
+
+## Rows
+These are the output rows of the View's Query. Rows are **not required**, Views can contain 0 rows.
+
+## Filters
+Views can have filters applied. Unlike Tables, view filters are not necessarily related to the columns that are present in the view.
+
+Using the example table above, imagine a view created from the query `SELECT ID, Title FROM Movies WHERE Year > 2000;` This will return this view: which is filtered by Year even though it's not a column in the View.
```

Comment:
Do we intend to show this filter in the 'filters dropdown' on the frontend?

I think it's best if we consider this just as part of the query for the view, and not as a filter that can be manipulated.

Reasoning:
* The view is specifically created by applying this filter on the parent table.
* When the user wants to apply filters on the view, I assume they'll want to only further filter the view with the columns shown rather than being able to edit the value for Year.
* Based on our UX, it would not be possible to add this filter for Year back if the user removes it, unless they close and reopen the view.
* It would also confuse non-technical users if a random filter is present on a view which does not contain that column.

Having said that, if we do not intend to show it on the UI, I think it's best to explicitly mention it here.

Reply:
> the 'filters dropdown' on the frontend?
> [...]
> Based on our UX

Please don't assume any particular design for Views as yet. This spec is aimed at clarifying product requirements which will then influence the requirements for design.

Reply:
e.g. see this UX for adding a filter in Chart.io https://chartio.com/docs/visual-sql/start-a-query/visual-mode/#add-a-filter.

Even if we don't allow the user to manipulate the filters, it seems like it would be useful to show the filter applied (without them having to look at the query or understand SQL), otherwise non-technical users may get confused about why the View is not showing all the data they expect. And if we're figuring out how to represent the filter, that's most of the work, so why not let them change it?

Reply:
Okay, that makes sense.

I've been explaining to people lately that tables are a structured representation of raw data and views are reports generated from those tables. When I look at views as reports, it kind of makes them seem immutable, which is true from a db standpoint.

But considering that we intend to represent views as mutable entities on the frontend, allowing users to edit the base filter makes sense.

I'm only concerned on how well we would be able to represent the differences between tables and views to the user.

Reply:
Argh, I couldn't see this conversation while doing my own review, so there's some overlap.  To reiterate what I said in other spots, I think we're missing clarity between the filter that defines a view, and a filter applied to a view.  I think we need both, and since from SQLAlchemy a view is a table (more-or-less), filtering a previously created view is the same as filtering a table.  I think we should be crystal clear on this point, lest we create confusion for users.

This applies to all transformations that can be used to define a view.

Reply:
I'm not sure that helps.  For me most natural would be:
- fiddle with a table (or tables) to get things set up like I want, then
- save that as a view, then
- recurse (i.e., do the same thing, but on the view).

When it comes to all the interactions that create a view (filtering, grouping, joining, choosing columns, etc.), a view has (at the DB level) precisely the same functionality as a table.  I think reflecting that in the UI is really empowering, since it lets a user iterate over different views to get what they want instead of needing to foresee everything when initially creating the view.

Reply:
The problem, then, is how to show the previous filter (or sequence of filtration steps) in a user-friendly way.  It'd be visible in the query if you know what you're looking for, but that won't be very friendly to non-technical types.

Reply:
I think that rather than having two different types of "filters", it would be better to just have one concept of filters and either portray them as "saved" or "unsaved". 

"Saved" filters are part of the query. "Unsaved" filters are only visible in the UI.

Reply:
> "Saved" filters are part of the query. "Unsaved" filters are only visible in the UI.

This would get rid of concept of the views and users would just see see it as "Saved filters" and can be misleading, if we happen to cache or store the filters locally. Moreover it might be confusing for views that are very deep. 
`View name` should be enough to know about the data that belongs to the view. For example, `views` for the table `Movies` would be named as "Horror Movies" enough to know what data to expect.

 I would rather have a dependency chart that would tell me the information how the table was derived than have a cluttered filters

Reply:
> This would get rid of concept of the views.

I don't see how this would get rid of the concept of views. Could you elaborate?

> users would just see see it as "Saved filters"

Views _are_ saved filters, among other things. Views can be used as:
- a way to save filters/sorts of a single table.
- a way to see data from multiple tables joined together.
- a way to aggregate data (either across a single table or multiple tables)
- a way to have computed/derived columns

...and probably more.

> `View name` should be enough to know about the data that belongs to the view. For example, `views` for the table `Movies` would be named as "Horror Movies" enough to know what data to expect.

I don't think we can rely on users always naming views well. I'm also not sure how this point relates to the product spec or filters.

> I would rather have a dependency chart that would tell me the information how the table was derived than have a cluttered filters

I'm not sure what you mean here by "a dependency chart" or "cluttered filters". Could you explain or maybe do a quick wireframe of what you're imagining for both of those things?

Reply:
The problem with just having saved filters is that they compose, and the composition order can be important (e.g., for "is distinct" filtering).  So, a view might be a filtered version of a join of two views, each of which is formed by filtering some table or another.  I think any UI that attempts to show the arbitrary composition of filters (or other operations used to create a view) is going to have the potential to become pretty cluttered pretty quickly.  I don't know if a dependency graph is the right solution, but I think it's crucial that a user understands when they're applying a filter to a view as opposed to an underlying view or an underlying data set.  I don't think "saved filter" gets that point across.

Reply:
> >"Saved" filters are part of the query. "Unsaved" filters are only visible in the UI.

> I don't see how this would get rid of the concept of views. Could you elaborate?

Views are more than just saved filters. They can be referenced by other queries/views,  can have their own set of filters/sorts, more of hierarchical structures which are closer to a table in characteristics. If we call them unsaved filters, it would be misleading as `unsaved filter` looks like something that cannot be extended up, something that either can be persisted or can't be and if persisted can be concatenated to existing filters
In the case of very deep Views(table1-> view1->view2), they need to show that hierarchy and users should be able to determine where certain operations like filter/sort took place.

Reply:
> Views are saved filters, among other things. Views can be used as:

Hierarchy and composition are the reason why I would not call them saved filters.

Reply:
@silentninja 

> Hierarchy and composition are the reason why I would not call them saved filters.

This is a succinct way to put it; I agree.

Reply:
> I don't think we can rely on users always naming views well. I'm also not sure how this point relates to the product spec or filters.

Views should be treated as similar to a table. so a view name could help with understanding what the data the view holds instead of having to look where or how it was generated. This is more of a convention rather an accurate description.  If named properly, we don't have to worry about showing how the view was generated unless required, where we could do complex diagrams like dependency graphs.

 Generally, when working with a `View` I wouldn't worry about how it was generated, rather would be focusing on what to do with the data it holds. This again is a convention, so it wouldn't apply to all the situations just like you mentioned

Reply:
> I'm not sure what you mean here by "a dependency chart" or "cluttered filters". Could you explain or maybe do a quick wireframe of what you're imagining for both of those things?

Dependency graph: 
![img](https://user-images.githubusercontent.com/4469754/150741800-a81db3c4-d0c3-488c-8820-c6d090f16fd5.png)

As for cluttered filters, I don't have visualisation other than our existing filter dropdown, but the reason why I think it would be cluttered is that we could end up with too many filters as a list(maybe along with the entity name they came from) in case of a deep view